### PR TITLE
Mediensteuerung

### PR DESCRIPTION
**Pull request by mastermichi**
First try to show and control notifications with mediastyle

### CODE REVIEW

File: src/com/firebirdberlin/nightdream/NotificationReceiver.java [hunk 1]
```diff
@@ -36,66 +49,164 @@ public void setColor(int color) {
...
+
         if (intent == null || context == null) return;
         if (Utility.isDebuggable(context)){
             Log.d(TAG, "Broadcast received.");
             dumpIntent(intent);
         }
+        SharedPreferences preferences = context.getSharedPreferences(PREFS_KEY, 0);
+
+        boolean showNotification = preferences.getBoolean("showNotification", true);
+        Log.d(TAG, "showNotification: "+showNotification);
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/nightdream/NotificationReceiver.java:73]: Usually I am trying to encapsulate calls to the SharedPreferences in Settings.java

e.g. like in `useNotificationStatusBar`

File: src/com/firebirdberlin/nightdream/NotificationReceiver.java [hunk 2]
```diff
@@ -36,66 +49,164 @@ public void setColor(int color) {
...
+                layoutParams.setMaxWidth(Utility.dpToPx(context, 24.f));
+                layoutParams.setMaxHeight(Utility.dpToPx(context, 24.f));
+                image.setLayoutParams(layoutParams);
+                container.addView(image);
+
+                //MediaStyle Control
+
+                FlexboxLayout mediastylecontrolcontainer = contentView.findViewById(R.id.notification_mediacontrol_bar);
+
+                boolean showmediastyle = preferences.getBoolean("showMediaStyle", false);
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/nightdream/NotificationReceiver.java:124]: direct access to preferences. You'd better encapsulate it. 

`showMediaStyle` would improve readability

File: src/com/firebirdberlin/nightdream/NotificationReceiver.java [hunk 3]
```diff
@@ -36,66 +49,164 @@ public void setColor(int color) {
...
+                layoutParams.setFlexGrow(0.f);
+                layoutParams.setHeight(Utility.dpToPx(context, 24.f));
+                layoutParams.setMaxWidth(Utility.dpToPx(context, 24.f));
+                layoutParams.setMaxHeight(Utility.dpToPx(context, 24.f));
+                image.setLayoutParams(layoutParams);
+                container.addView(image);
+
+                //MediaStyle Control
+
+                FlexboxLayout mediastylecontrolcontainer = contentView.findViewById(R.id.notification_mediacontrol_bar);
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/nightdream/NotificationReceiver.java:122]: Please use lower camelCase: `mediaStyleControlContainer`

File: src/com/firebirdberlin/nightdream/NotificationReceiver.java [hunk 4]
```diff
@@ -36,66 +49,164 @@ public void setColor(int color) {
...
+                    mediastylecontrolcontainer.removeAllViews();
+
+                    LayoutInflater inflater = (LayoutInflater)
+                            context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
+
+                    View mediastylecontrol = inflater.inflate(R.layout.notification_mediacontrol, null);
+                    mediastylecontrol.findViewById(R.id.notify_largeicon).setVisibility(View.GONE);
+                    mediastylecontrol.findViewById(R.id.notify_largepicture).setVisibility(View.GONE);
+
+                    ImageView notificationMessagebitmap = (ImageView) mediastylecontrol.findViewById(R.id.notify_smallicon);
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/nightdream/NotificationReceiver.java:139]: cast to (ImageView) is redundant

File: src/com/firebirdberlin/nightdream/NotificationReceiver.java [hunk 5]
```diff
@@ -36,66 +49,164 @@ public void setColor(int color) {
...
+                    mediastylecontrol.findViewById(R.id.notify_actiontext1).setVisibility(View.GONE);
+                    mediastylecontrol.findViewById(R.id.notify_actiontext2).setVisibility(View.GONE);
+                    mediastylecontrol.findViewById(R.id.notify_actiontext3).setVisibility(View.GONE);
+
+                    ImageView[] notificationActionImages = new ImageView[5];
+                    notificationActionImages[0] = (ImageView) mediastylecontrol.findViewById(R.id.notify_actionview1);
+                    notificationActionImages[1] = (ImageView) mediastylecontrol.findViewById(R.id.notify_actionview2);
+                    notificationActionImages[2] = (ImageView) mediastylecontrol.findViewById(R.id.notify_actionview3);
+                    notificationActionImages[3] = (ImageView) mediastylecontrol.findViewById(R.id.notify_actionview4);
+                    notificationActionImages[4] = (ImageView) mediastylecontrol.findViewById(R.id.notify_actionview5);
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/nightdream/NotificationReceiver.java:168]: cast is redundant

File: src/com/firebirdberlin/nightdream/NotificationReceiver.java [hunk 6]
```diff
@@ -36,66 +49,164 @@ public void setColor(int color) {
...
+                container.addView(image);
+
+                //MediaStyle Control
+
+                FlexboxLayout mediastylecontrolcontainer = contentView.findViewById(R.id.notification_mediacontrol_bar);
+
+                boolean showmediastyle = preferences.getBoolean("showMediaStyle", false);
+
+                if (Objects.requireNonNull(intent.getStringExtra("template")).contains("MediaStyle") && showmediastyle) {
+                    Log.w(TAG, "Show MediaStyle notification");
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/nightdream/NotificationReceiver.java:127]: probably it makes sense to put this entire block into an own function

File: src/com/firebirdberlin/nightdream/NightDreamActivity.java [hunk 7]
```diff
@@ -226,6 +228,8 @@ public void onCreate(Bundle savedInstanceState) {
         weatherIcon = findViewById(R.id.icon_weather_forecast);
         alarmClockIcon = findViewById(R.id.alarm_clock_icon);
         radioIcon = findViewById(R.id.radio_icon);
+
+
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/nightdream/NightDreamActivity.java:232]: unnecessary modification

File: src/com/firebirdberlin/nightdream/PreferencesFragment.java [hunk 8]
```diff
@@ -288,26 +297,71 @@ public void onActivityCreated(Bundle savedInstanceState) {
...
+                    setPreferencesFromResource(R.xml.preferences_brightness, rootKey);
+                    break;
+                case "behaviour":
+                    setPreferencesFromResource(R.xml.preferences_behaviour, rootKey);
+                    break;
+                case "nightmode":
+                    setPreferencesFromResource(R.xml.preferences_nightmode, rootKey);
+                    break;
+                case "notifications":
+                    setPreferencesFromResource(R.xml.preferences_notifications, rootKey);
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/nightdream/PreferencesFragment.java:327]: This block should be moved to a separate function.

File: src/com/firebirdberlin/nightdream/PreferencesFragment.java [hunk 9]
```diff
@@ -1029,4 +1083,49 @@ public void onClick(View v) {
...
+                }
+            }
+
+            if(started) {
+                start_notifyaccess_changed();
+            }
+        }
+    };
+
+    public void stop_notifyaccess_changed() {
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/nightdream/PreferencesFragment.java:1122]: currently unused

File: src/com/firebirdberlin/nightdream/PreferencesFragment.java [hunk 10]
```diff
@@ -1029,4 +1083,49 @@ public void onClick(View v) {
...
+        started = false;
+        handler.removeCallbacks(notifyinformation_changed);
+    }
+
+    public void start_notifyaccess_changed() {
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/nightdream/PreferencesFragment.java:1127]: camelCase

File: src/com/firebirdberlin/nightdream/PreferencesFragment.java [hunk 11]
```diff
@@ -53,6 +55,9 @@
     private static int RESULT_LOAD_IMAGE_KITKAT = 4;
     private final int PERMISSIONS_REQUEST_READ_EXTERNAL_STORAGE = 2;
     private final int PERMISSIONS_REQUEST_RECORD_AUDIO = 3;
+    private Handler handler = new Handler();
+    private boolean started;
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/nightdream/PreferencesFragment.java:59]: `started` is way too generic. A more speaking name would be helpful.

File: src/com/firebirdberlin/nightdream/PreferencesFragment.java [hunk 12]
```diff
@@ -97,7 +103,10 @@ public void run() {
                         }
 
                     }
+
                     switch (key) {
+                        case "showNotification":
+                            break;
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/nightdream/PreferencesFragment.java:109]: why is this necessary ?

File: src/com/firebirdberlin/nightdream/PreferencesFragment.java [hunk 13]
```diff
@@ -1029,4 +1083,49 @@ public void onClick(View v) {
             }
         }
     }
+
+    private Runnable notifyinformation_changed = new Runnable() {
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/nightdream/PreferencesFragment.java:1087]: camelCase

File: src/com/firebirdberlin/nightdream/PreferencesFragment.java [hunk 14]
```diff
@@ -1029,4 +1083,49 @@ public void onClick(View v) {
...
+                    try {
+                        numberPreference.setSummary(getString(R.string.showNotificationsaccess));
+                        numberPreference.setEnabled(true);
+                    }catch (Exception ex){
+
+                    }
+                }
+            }
+
+            if(started) {
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/nightdream/PreferencesFragment.java:1116]: restart every 2s ? Should this be `stop_notifyaccess_changed` ?

File: src/com/firebirdberlin/nightdream/ui/NightDreamUI.java [hunk 15]
```diff
@@ -290,6 +290,7 @@ public boolean onSingleTapUp(MotionEvent e) {
             return false;
         }
 
+
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/nightdream/ui/NightDreamUI.java:293]:
```suggestion
```

**Comment by mastermichi** [comment]
> Introducing a new preference 'showNotifications' with default False will disable notifications for most users when they update. What about setting the default to true ?

Setting to "false" is the standard because the android setting to read notifications maybe has not been set yet.
But the update is a good point I did not thought of. Maybe it the standard should be "true". No objections because it is disabled in the setup if there is no right for NachtUhr to read it.

File: res/layout-v17/main.xml [hunk 16]
```diff
@@ -52,21 +52,22 @@
...
                     android:id="@+id/notificationstatusbar"
                     android:layout_width="0dp"
                     android:layout_height="match_parent"
                     android:layout_weight="1"
-                    app:flexDirection="row"
                     app:alignContent="flex_end"
-                    app:justifyContent="flex_end">
-                </com.google.android.flexbox.FlexboxLayout>
+                    app:flexDirection="row"
+                    app:justifyContent="flex_end"></com.google.android.flexbox.FlexboxLayout>
```
[the remark this answers is not in the document]
Comment (mastermichi) [reply res/layout-v17/main.xml:67]: Just Layout. Change it if you like.

File: res/layout-v17/main.xml [hunk 17]
```diff
@@ -52,21 +52,22 @@
...
+
```
[the remark this answers is not in the document]
Comment (mastermichi) [reply res/layout-v17/main.xml:68]: Just Layout. Change it if you like.

File: res/layout-v17/main.xml [hunk 18]
```diff
@@ -52,21 +52,22 @@
...
             </LinearLayout>
+
```
[the remark this answers is not in the document]
Comment (mastermichi) [reply res/layout-v17/main.xml:70]: Just Layout. Change it if you like.

File: res/layout-v17/main.xml [hunk 19]
```diff
@@ -101,6 +102,8 @@
         <com.firebirdberlin.nightdream.ui.BottomPanelLayout
             android:id="@+id/bottomPanel"
             android:layout_width="match_parent"
-            android:layout_height="60dp"/>
+            android:layout_height="60dp">
+
+        </com.firebirdberlin.nightdream.ui.BottomPanelLayout>
```
[the remark this answers is not in the document]
Comment (mastermichi) [reply res/layout-v17/main.xml:107]: Just Layout. Change it if you like.

File: res/layout/notification_mediacontrol.xml [hunk 20]
```diff
@@ -0,0 +1,293 @@
...
+
+            android:minWidth="300dp"
+            app:layout_constraintEnd_toEndOf="parent"
+            app:layout_constraintStart_toEndOf="parent"
+            app:layout_constraintStart_toStartOf="parent"
+            app:layout_constraintTop_toTopOf="parent">
+
+            <LinearLayout
+                android:id="@+id/notify_background2"
+                android:layout_width="fill_parent"
```
[the remark this answers is not in the document]
Comment (mastermichi) [reply res/layout/notification_mediacontrol.xml:38]: Just Layout. Change it if you like.

File: src/com/firebirdberlin/nightdream/PreferencesFragment.java (hunk 11 above)
Comment (mastermichi) [reply src/com/firebirdberlin/nightdream/PreferencesFragment.java:59]: No problem. Change it to a speaking name you like.

File: src/com/firebirdberlin/nightdream/PreferencesFragment.java (hunk 12 above)
Comment (mastermichi) [reply src/com/firebirdberlin/nightdream/PreferencesFragment.java:109]: Just implemented for feature releases. Got a new idea for a feature (a feature I can not remember anymore :-( ) but I have not found the time to implement it. You can delete it if you like.

File: src/com/firebirdberlin/nightdream/PreferencesFragment.java (hunk 13 above)
Comment (mastermichi) [reply src/com/firebirdberlin/nightdream/PreferencesFragment.java:1087]: No problem. Change it if you like.

File: src/com/firebirdberlin/nightdream/ui/NightDreamUI.java (hunk 15 above)
Comment (mastermichi) [reply src/com/firebirdberlin/nightdream/ui/NightDreamUI.java:293]: No problem. Change it if you like.

File: src/com/firebirdberlin/nightdream/PreferencesFragment.java (hunk 14 above)
Comment (mastermichi) [reply src/com/firebirdberlin/nightdream/PreferencesFragment.java:1116]: It is necessary for feature releases. If you go to Android Setup from NightDream and change the status of the notification right it will not be shown in NightDream until you refresh the display. Now you go back from Android setup to the first page of NightDream setup, so you can not see any effect of this function. This should be fixed in feature releases so that the function "stop_notifyaccess_changed" is necessary. You sould go back from Android setup to NightDream notifications setup.

File: res/layout/notification_mediacontrol.xml [hunk 21]
```diff
@@ -0,0 +1,293 @@
...
+                android:layout_height="wrap_content"
+                android:orientation="vertical"
+                android:visibility="visible">
+
+                <LinearLayout
+                    android:id="@+id/notify_background"
+                    android:layout_width="fill_parent"
```
[the remark this answers is not in the document]
Comment (mastermichi) [reply res/layout/notification_mediacontrol.xml:45]: Just Layout. Change it if you like.

File: src/com/firebirdberlin/nightdream/BillingHelperActivity.java [hunk 22]
```diff
@@ -65,7 +65,7 @@
     SharedPreferences preferences;
 
     public boolean isPurchased(String sku) {
-        if (Utility.isEmulator()) {
+        if (!Utility.isEmulator()) {
```
[the remark this answers is not in the document]
Comment (mastermichi) [reply src/com/firebirdberlin/nightdream/BillingHelperActivity.java:68]: Sorry

File: src/com/firebirdberlin/nightdream/NotificationReceiver.java (hunk 2 above)
Comment (mastermichi) [reply src/com/firebirdberlin/nightdream/NotificationReceiver.java:124]: OK

File: src/com/firebirdberlin/nightdream/NotificationReceiver.java (hunk 1 above)
Comment (mastermichi) [reply src/com/firebirdberlin/nightdream/NotificationReceiver.java:73]: ok

File: src/com/firebirdberlin/nightdream/PreferencesFragment.java (hunk 9 above)
Comment (mastermichi) [reply src/com/firebirdberlin/nightdream/PreferencesFragment.java:1122]: Implemented for future release

File: src/com/firebirdberlin/nightdream/PreferencesFragment.java (hunk 10 above)
Comment (mastermichi) [reply src/com/firebirdberlin/nightdream/PreferencesFragment.java:1127]: let's change it

File: src/com/firebirdberlin/nightdream/PreferencesFragment.java (hunk 8 above)
Comment (mastermichi) [reply src/com/firebirdberlin/nightdream/PreferencesFragment.java:327]: ok. let's do it.

File: src/com/firebirdberlin/nightdream/NightDreamActivity.java (hunk 7 above)
Comment (mastermichi) [reply src/com/firebirdberlin/nightdream/NightDreamActivity.java:232]: ok

File: src/com/firebirdberlin/nightdream/NotificationReceiver.java (hunk 6 above)
Comment (mastermichi) [reply src/com/firebirdberlin/nightdream/NotificationReceiver.java:127]: ok. let's do it.

File: src/com/firebirdberlin/nightdream/NotificationReceiver.java (hunk 5 above)
Comment (mastermichi) [reply src/com/firebirdberlin/nightdream/NotificationReceiver.java:168]: If you got a better solution make it. Maybe some old code.

File: src/com/firebirdberlin/nightdream/NotificationReceiver.java (hunk 4 above)
Comment (mastermichi) [reply src/com/firebirdberlin/nightdream/NotificationReceiver.java:139]: is redundant but better to read the code

File: src/com/firebirdberlin/nightdream/NotificationReceiver.java (hunk 3 above)
Comment (mastermichi) [reply src/com/firebirdberlin/nightdream/NotificationReceiver.java:122]: No problem. let's change it.

File: src/com/firebirdberlin/nightdream/NightDreamActivity.java [hunk 23]
```diff
@@ -25,6 +25,7 @@
 import android.os.PowerManager;
 import android.util.Log;
 import android.view.GestureDetector;
+import android.view.LayoutInflater;
```
[the remark this answers is not in the document]
Comment (mastermichi) [reply src/com/firebirdberlin/nightdream/NightDreamActivity.java:28]: kill it. Maybe some old code.

File: src/com/firebirdberlin/nightdream/NightDreamActivity.java [hunk 24]
```diff
@@ -33,6 +34,7 @@
 import android.widget.ImageView;
 import android.widget.Toast;
 
+import androidx.annotation.ContentView;
```
[the remark this answers is not in the document]
Comment (mastermichi) [reply src/com/firebirdberlin/nightdream/NightDreamActivity.java:37]: Kill it. Maybe some old code.

**Comment by firebirdberlin** [comment]
Closed. will be continued in a new PR.
